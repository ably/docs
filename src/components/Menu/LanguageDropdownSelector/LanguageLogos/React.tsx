import React from 'react';

// The best one.
export const ReactJS = () => (
  <svg width="24" height="24" viewBox="-11.5 -10.23174 23 20.46348" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="0" cy="0" r="2.05" fill="#61dafb" />
    <g stroke="#61dafb" strokeWidth="1" fill="none">
      <ellipse rx="11" ry="4.2" />
      <ellipse rx="11" ry="4.2" transform="rotate(60)" />
      <ellipse rx="11" ry="4.2" transform="rotate(120)" />
    </g>
  </svg>
);
