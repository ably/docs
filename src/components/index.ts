export * from './ButtonWithTooltip';
export * from './LanguageButton';
export * from './HorizontalMenu';
