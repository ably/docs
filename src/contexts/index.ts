export * from './page-language-context';
export { default as PageLanguageContext } from './page-language-context';
export * from './user-context';
