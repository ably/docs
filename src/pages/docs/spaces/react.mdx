---
title: React Hooks
meta_description: "Incorporate Spaces into your React application with idiomatic and user-friendly React Hooks."
---

Incorporate Spaces into your React application with idiomatic and user-friendly React Hooks.

This package enables you to:

* Interact with Ably [Spaces](/docs/spaces) using a React Hook
* Subscribe to [events](/docs/spaces/avatar#events) in a space
* Retrieve the [membership](/docs/spaces/space) of a space
* Set the [location](/docs/spaces/locations) of space members
* Acquire [locks](/docs/spaces/locking) on components within a space
* Set the position of [members' cursors](/docs/spaces/cursors) in a space

The following hooks are available:

| Hook | Description |
| ---- | ----------- |
| [`useSpace`](#useSpace) | The `useSpace` hook lets you subscribe to the current Space, receive Space state events, and get the current Space instance. |
| [`useMembers`](#useMembers) | The `useMembers` hook is useful in building avatar stacks. Using `useMembers`, you can retrieve space members. |
| [`useLocations`](#useLocations) | The `useLocations` hook lets you subscribe to location events. Location events are emitted whenever a member changes location. |
| [`useLocks`](#useLocks) | The `useLocks` hook lets you subscribe to lock events by registering a listener. Lock events are emitted whenever the lock state transitions into `locked` or `unlocked`. |
| [`useLock`](#useLock) | The `useLock` hook returns the status of a lock and, if it has been acquired, the member holding the lock. |
| [`useCursors`](#useCursors) | The `useCursors` hook allows you to track a member's pointer position updates across an application. |

Spaces hooks are client-side oriented. If you employ server-side rendering, ensure components using these hooks render only on the client-side.

## Install <a id="install"/>

<Code>
```shell
npm install ably @ably/spaces
```
</Code>

<Aside data-type='note'>
React version 16.8.0 or above is required.
</Aside>

## Authenticate <a id="authenticate"/>

An [API key](/docs/auth#api-keys) is required to authenticate with Ably. API keys are used either to authenticate directly with Ably using [basic authentication](/docs/auth/basic), or to generate tokens for untrusted clients using [token authentication](/docs/auth/token).

[Sign up](https://ably.com/sign-up) to Ably to create an API key in the [dashboard](https://ably.com/dashboard) or use the [Control API](/docs/platform/account/control-api) to create an API programmatically.

<Aside data-type='important'>
The examples use [basic authentication](/docs/auth/basic) to demonstrate usage for convenience. In your own applications, basic authentication should never be used on the client-side as it exposes your Ably API key. Instead use [token authentication](/docs/auth/token).
</Aside>

## Usage <a id="usage"/>

### Setting up the Spaces Provider <a id="spaces-provider"/>

Use the `SpacesProvider` component to connect to Ably. The `SpacesProvider` should wrap every component that needs to access Spaces.

<Code>
```react
import { Realtime } from "ably";
import Spaces from "@ably/spaces";
import { SpacesProvider, SpaceProvider } from "@ably/spaces/react";

const ably = new Realtime({ key: "{{API_KEY}}", clientId: 'clemons' });
const spaces = new Spaces(ably);

root.render(
  <SpacesProvider client={spaces}>
    <SpaceProvider name="my-space">
      <App />
    </SpaceProvider>
  </SpacesProvider>
)
```
</Code>

### useSpace <a id="useSpace"/>

The `useSpace` hook enables you to subscribe to the current [Space](/docs/spaces/space), receive Space state events, and get the current Space instance.

<Code>
```react
const { space } = useSpace((update) => {
  console.log(update);
});
```
</Code>

### useMembers <a id="useMembers"/>

`useMembers` is used to build [avatar stacks](/docs/spaces/avatar). It retrieves members of the space, including members that have recently left the space, but have not yet been removed.

<Code>
```react
const { self, others, members } = useMembers();
```
</Code>

- `self` - a memberâ€™s own member object
- `others` - an array of member objects for all members other than the member themselves
- `members` - an array of all member objects, including the member themselves

`useMembers` also enables you to subscribe to members entering, leaving, and being removed from the Space (after a timeout), as well as when a member updates their [profile information](/docs/spaces/space#update-profile).

<Code>
```react
// Subscribe to all member events in a space
useMembers((memberUpdate) => {
  console.log(memberUpdate);
});

// Subscribe to member enter events only
useMembers('enter', (memberJoined) => {
  console.log(memberJoined);
});

// Subscribe to member leave events only
useMembers('leave', (memberLeft) => {
  console.log(memberLeft);
});

// Subscribe to member remove events only
useMembers('remove', (memberRemoved) => {
  console.log(memberRemoved);
});

// Subscribe to profile updates on members only
useMembers('updateProfile', (memberProfileUpdated) => {
  console.log(memberProfileUpdated);
});

// Subscribe to all updates to members
useMembers('update', (memberUpdate) => {
  console.log(memberUpdate);
});
```
</Code>

### useLocations <a id="useLocations"/>

`useLocations` enables you to subscribe to [location](/docs/spaces/locations) events. Location events are emitted whenever a member changes location.

<Code>
```react
useLocations((locationUpdate) => {
  console.log(locationUpdate);
});
```
</Code>

`useLocations` also enables you to update the current member location using the `update` method provided by the hook. For example:

<Code>
```react
const { update } = useLocations((locationUpdate) => {
  console.log(locationUpdate);
});
```
</Code>

### useLocks <a id="useLocks"/>

`useLocks` enables you to subscribe to [lock](/docs/spaces/locking) events by registering a listener. Lock events are emitted whenever a lock transitions into the `locked` or `unlocked` state.

<Code>
```react
useLocks((lockUpdate) => {
  console.log(lockUpdate);
});
```
</Code>

### useLock <a id="useLock"/>

`useLock` returns the status of a [lock](/docs/spaces/locking) and, if the lock has been acquired, the member holding that lock.

<Code>
```react
const { status, member } = useLock('my-lock');
```
</Code>

### useCursors <a id="useCursors"/>

`useCursors` enables you to track a member's [cursor](/docs/spaces/cursors) position and provide a view of all members' cursors within a space. For example:

<Code>
```react
// Subscribe to events published on "mousemove" by all members
const { set } =  useCursors((cursorUpdate) => {
  console.log(cursorUpdate);
});

useEffect(() => {
  // Publish a your cursor position on "mousemove" including optional data
  const eventListener = ({ clientX, clientY }) => {
    set({ position: { x: clientX, y: clientY }, data: { color: 'red' } });
  }

  window.addEventListener('mousemove', eventListener);

  return () => {
    window.removeEventListener('mousemove', eventListener);
  };
});
```
</Code>

If you provide `{ returnCursors: true }` as an option you can retrieve active members' cursors:

<Code>
```react
const { cursors } =  useCursors((cursorUpdate) => {
  console.log(cursorUpdate);
}, { returnCursors: true });
```
</Code>

### Error handling <a id="error-handling"/>

[`useSpace`](#useSpace), [`useMembers`](#useMembers), [`useLocks`](#useLocks), and [`useCursors`](#useCursors) return [connection](/docs/connect) and [channel](/docs/channels) errors you may encounter, so that you can handle them within your components. This may include when a client doesn't have permission to attach to a channel, or if it loses its connection to Ably.

<Code>
```react
const { connectionError, channelError } = useMembers();

if (connectionError) {
  // TODO: handle connection errors
} else if (channelError) {
  // TODO: handle channel errors
} else {
  return <SpacesPoweredComponent />
}
```
</Code>
