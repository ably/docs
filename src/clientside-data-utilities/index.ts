import { isString } from 'lodash/fp';
import { HtmlComponentPropsData } from 'src/components/html-component-props';

export const attribsContainClass = (
  cssClass: string,
  attribs: {
    className?: string;
  },
) =>
  attribs.className &&
  (attribs.className === cssClass ||
    attribs.className.includes(` ${cssClass} `) ||
    attribs.className.substring(0, cssClass.length + 1) === `${cssClass} ` ||
    attribs.className.substring(attribs.className.length - (cssClass.length + 1)) === ` ${cssClass}`);

export const ensureChildDataShows = (data: HtmlComponentPropsData): HtmlComponentPropsData => {
  if (!data) {
    return data;
  }
  if (isString(data)) {
    return data;
  }
  return data.map((child) => ({
    ...child,
    data: ensureChildDataShows(child.data),
    attribs: { ...child.attribs, forcedisplay: 'true' },
  }));
};
