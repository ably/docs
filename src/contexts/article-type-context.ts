import { ARTICLE_TYPES } from '../../data/transform/constants';
import { createContext } from 'react';

type ArticleType = keyof typeof ARTICLE_TYPES;

export const ArticleTypeContext = createContext<ArticleType>(ARTICLE_TYPES.document);
