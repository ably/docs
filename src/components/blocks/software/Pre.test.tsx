import React from 'react';
import { render, screen } from '@testing-library/react';
import { PageLanguageContext } from 'src/contexts';
import { Pre } from './';

describe('<Pre />', () => {
  it('should successfully render Code elements with language', () => {
    const { container } = render(
      <Pre
        data={[{ data: 'var useVars = false;', type: 'text' }]}
        language="javascript"
        languages={['javascript', 'csharp']}
        altData={{
          javascript: [{ data: 'var useVars = false;', type: 'text' }],
          csharp: [{ data: 'bool useVars = false;', type: 'text' }],
        }}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render tip wrapper if language is unavailable', () => {
    render(
      <PageLanguageContext.Provider value="javascript">
        <Pre
          data={[{ data: 'useVars = false;', type: 'text' }]}
          language="javascript"
          languages={['python', 'csharp']}
          altData={{
            python: [{ data: 'useVars = false;', type: 'text' }],
            csharp: [{ data: 'bool useVars = false;', type: 'text' }],
          }}
        />
      </PageLanguageContext.Provider>,
    );
    expect(screen.getByText('we donâ€™t yet have a relevant code sample', { exact: false })).toBeInTheDocument();
  });

  it('should render code block when no languages are passed', () => {
    const { container } = render(
      <Pre data={[{ data: 'var useVars = false;', type: 'text' }]} language="javascript" languages={undefined} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
