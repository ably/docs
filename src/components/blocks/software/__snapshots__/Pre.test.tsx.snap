// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pre /> should render code block when no languages are passed 1`] = `
<div
  class="my-32"
>
  <pre
    class="bg-cool-black text-white p-0 rounded-lg relative"
  >
    useVars = false;
  </pre>
</div>
`;

exports[`<Pre /> should successfully render Code elements with language 1`] = `
<div
  class="my-32"
>
  <pre
    class="bg-cool-black text-white p-0 rounded-lg relative"
  >
    <div
      class="border-b border-charcoal-grey w-full"
    >
      <menu
        class="horizontalNav"
        data-testid="menu"
      >
        <button
          class="button isActive"
        >
          JavaScript
        </button>
        <button
          class="button"
        >
          C# .NET
        </button>
        <div
          class="dropdownContainer"
        >
          <div
            class=" css-fyq6mk-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-language-navigation-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="ably-select__control css-bo7mvc-control"
            >
              <div
                class="ably-select__value-container ably-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <div
                  class="ably-select__single-value css-ue700n-singleValue"
                >
                  JavaScript v2.0
                </div>
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-language-navigation-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="ably-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="ably-select__indicator-separator css-1xjgjl1-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="ably-select__indicator ably-select__dropdown-indicator css-1gvtn5f-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
    useVars = false;
  </pre>
</div>
`;

exports[`<Pre /> should successfully render code elements with  only Realtime languages 1`] = `
<div
  class="my-32"
>
  <pre
    class="bg-cool-black text-white p-0 rounded-lg relative"
  >
    <div
      class="bg-dark-grey border-charcoal-grey text-white border-b-4 rounded-t-lg flex justify-end"
    >
      <menu
        class="flex md:overflow-x-auto pl-0 justify-end md:justify-start h-48 mr-16 my-0"
        data-testid="menuSDKInterface"
      >
        <button
          class="font-medium font-sans focus:outline-none px-24 text-mid-grey bg-charcoal-grey"
        >
          Realtime
        </button>
        <div
          aria-label="Ably SDKs may contain a realtime and a REST interface, each of which satisfy different use cases."
          class="flex flex-row w-full justify-start mt-2"
          role="button"
          tabindex="0"
        >
          <svg
            class="mid-grey mt-12 ml-16"
            style="width: 1.25rem; height: 1.25rem;"
          >
            <use
              xlink:href="#sprite-icon-gui-info"
            />
          </svg>
        </div>
      </menu>
    </div>
    <div
      class="border-b border-charcoal-grey w-full"
    >
      <menu
        class="horizontalNav"
        data-testid="menu"
      >
        <button
          class="button isActive"
        >
          JavaScript
        </button>
        <button
          class="button"
        >
          C# .NET
        </button>
        <div
          class="dropdownContainer"
        >
          <div
            class=" css-fyq6mk-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-language-navigation-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="ably-select__control css-bo7mvc-control"
            >
              <div
                class="ably-select__value-container ably-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <div
                  class="ably-select__single-value css-ue700n-singleValue"
                >
                  JavaScript v2.0
                </div>
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-language-navigation-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="ably-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="ably-select__indicator-separator css-1xjgjl1-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="ably-select__indicator ably-select__dropdown-indicator css-1gvtn5f-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
    useVars = false;
  </pre>
</div>
`;

exports[`<Pre /> should successfully render code elements with  only Rest languages 1`] = `
<div
  class="my-32"
>
  <pre
    class="bg-cool-black text-white p-0 rounded-lg relative"
  >
    <div
      class="bg-dark-grey border-charcoal-grey text-white border-b-4 rounded-t-lg flex justify-end"
    >
      <menu
        class="flex md:overflow-x-auto pl-0 justify-end md:justify-start h-48 mr-16 my-0"
        data-testid="menuSDKInterface"
      >
        <button
          class="font-medium font-sans focus:outline-none px-24 text-mid-grey "
        >
          Realtime
        </button>
        <div
          aria-label="Ably SDKs may contain a realtime and a REST interface, each of which satisfy different use cases."
          class="flex flex-row w-full justify-start mt-2"
          role="button"
          tabindex="0"
        >
          <svg
            class="mid-grey mt-12 ml-16"
            style="width: 1.25rem; height: 1.25rem;"
          >
            <use
              xlink:href="#sprite-icon-gui-info"
            />
          </svg>
        </div>
      </menu>
    </div>
    <div
      class="border-b border-charcoal-grey w-full"
    >
      <menu
        class="horizontalNav"
        data-testid="menu"
      >
        <button
          class="button isActive"
        >
          JavaScript
        </button>
        <button
          class="button"
        >
          C# .NET
        </button>
        <div
          class="dropdownContainer"
        >
          <div
            class=" css-fyq6mk-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-language-navigation-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="ably-select__control css-bo7mvc-control"
            >
              <div
                class="ably-select__value-container ably-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <div
                  class="ably-select__single-value css-ue700n-singleValue"
                >
                  JavaScript v2.0
                </div>
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-language-navigation-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="ably-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="ably-select__indicator-separator css-1xjgjl1-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="ably-select__indicator ably-select__dropdown-indicator css-1gvtn5f-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
    useVars = false;
  </pre>
</div>
`;

exports[`<Pre /> should successfully render code elements with both Rest and Realtime languages 1`] = `
<div
  class="my-32"
>
  <pre
    class="bg-cool-black text-white p-0 rounded-lg relative"
  >
    <div
      class="bg-dark-grey border-charcoal-grey text-white border-b-4 rounded-t-lg flex justify-end"
    >
      <menu
        class="flex md:overflow-x-auto pl-0 justify-end md:justify-start h-48 mr-16 my-0"
        data-testid="menuSDKInterface"
      >
        <button
          class="font-medium font-sans focus:outline-none px-24 text-mid-grey "
        >
          Realtime
        </button>
        <button
          class="font-medium font-sans focus:outline-none px-24 text-mid-grey bg-charcoal-grey"
        >
          REST
        </button>
        <div
          aria-label="Ably SDKs may contain a realtime and a REST interface, each of which satisfy different use cases."
          class="flex flex-row w-full justify-start mt-2"
          role="button"
          tabindex="0"
        >
          <svg
            class="mid-grey mt-12 ml-16"
            style="width: 1.25rem; height: 1.25rem;"
          >
            <use
              xlink:href="#sprite-icon-gui-info"
            />
          </svg>
        </div>
      </menu>
    </div>
    <div
      class="border-b border-charcoal-grey w-full"
    >
      <menu
        class="horizontalNav"
        data-testid="menu"
      >
        <button
          class="button isActive"
        >
          JavaScript
        </button>
        <button
          class="button"
        >
          C# .NET
        </button>
        <div
          class="dropdownContainer"
        >
          <div
            class=" css-fyq6mk-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-language-navigation-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="ably-select__control css-bo7mvc-control"
            >
              <div
                class="ably-select__value-container ably-select__value-container--has-value css-1fdsijx-ValueContainer"
              >
                <div
                  class="ably-select__single-value css-ue700n-singleValue"
                >
                  JavaScript v2.0
                </div>
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-readonly="true"
                  class="css-mohuvp-dummyInput-DummyInput"
                  id="react-select-language-navigation-input"
                  inputmode="none"
                  role="combobox"
                  tabindex="0"
                  value=""
                />
              </div>
              <div
                class="ably-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="ably-select__indicator-separator css-1xjgjl1-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="ably-select__indicator ably-select__dropdown-indicator css-1gvtn5f-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </menu>
    </div>
    useVars = false;
  </pre>
</div>
`;
